### PR TITLE
Implement Frost based DKG

### PR DESCRIPTION
**Pull request by gl-rockx**
- [x] Wrap Coinbase frost implementation
- [x] Add blame messages
- [ ] Adapt tests 
- [x] Support resharing
- [ ] More fine-grained error / blame handling
- [ ] Handle Timeout

### CODE REVIEW

File: dkg/frost/frost.go [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+
+var thisCurve = curves.BLS12381G1()
+
+func init() {
+	_ = bls.Init(bls.BLS12_381)
+	_ = bls.SetETHmode(bls.EthModeDraft07)
+}
+
+type FROST struct {
+	identifier   dkg.RequestID
```
Comment (alonmuroch) [inline dkg/frost/frost.go:26]: better have some kind of a "state" we can get a root from so we can use it for the spect tests

File: dkg/frost/frost.go [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+	"github.com/coinbase/kryptology/pkg/dkg/frost"
+	"github.com/coinbase/kryptology/pkg/sharing"
+	ecies "github.com/ecies/go/v2"
+	"github.com/herumi/bls-eth-go-binary/bls"
+	"github.com/pkg/errors"
+)
+
+var thisCurve = curves.BLS12381G1()
+
+func init() {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:20]: please use https://github.com/bloxapp/ssv-spec/blob/main/types/bls.go

File: dkg/frost/frost.go [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+	sessionSK  *ecies.PrivateKey
+
+	msgs           map[DKGRound]map[uint32]*dkg.SignedMessage
+	operatorShares map[uint32]*bls.SecretKey
+}
+
+type DKGRound int
+
+const (
+	Preparation DKGRound = iota + 1
```
Comment (alonmuroch) [inline dkg/frost/frost.go:45]: Why iota + 1?

File: dkg/frost/frost.go [hunk 4]
```diff
@@ -0,0 +1,756 @@
...
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+
+	"github.com/bloxapp/ssv-spec/dkg"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/coinbase/kryptology/pkg/dkg/frost"
+	"github.com/coinbase/kryptology/pkg/sharing"
+	ecies "github.com/ecies/go/v2"
```
Comment (alonmuroch) [inline dkg/frost/frost.go:13]: is this a well known and well used lib? I saw go-eth has an implementation as well

File: dkg/frost/frost.go [hunk 5]
```diff
@@ -0,0 +1,756 @@
...
+	identifier   dkg.RequestID
+	network      dkg.Network
+	signer       types.DKGSigner
+	storage      dkg.Storage
+	threshold    uint32
+	currentRound DKGRound
+
+	operatorID types.OperatorID
+	operators  []uint32
+	party      *frost.DkgParticipant
```
Comment (alonmuroch) [inline dkg/frost/frost.go:35]: party has a different meaning, better use participant or frostParticipant

File: dkg/frost/frost.go [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+	fr.currentRound = Preparation
+
+	protocolMessage := &ProtocolMsg{
+		Round: Preparation,
+		PreparationMessage: &PreparationMessage{
+			SessionPk: k.PublicKey.Bytes(true),
+		},
+	}
+
+	protocolMessageBytes, err := protocolMessage.Encode()
```
Comment (alonmuroch) [inline dkg/frost/frost.go:119]: L119-L150 is probably used elsewhere so why not make a reusable func for broadcasting?

File: dkg/frost/frost.go [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+	fr.msgs[Preparation][uint32(fr.operatorID)] = bcastPrepMessage
+
+	return fr.network.BroadcastDKGMessage(bcastPrepMessage)
+}
+
+func (fr *FROST) ProcessMsg(msg *dkg.SignedMessage) (bool, *dkg.KeyGenOutput, error) {
+	if err := msg.Validate(); err != nil {
+		return false, nil, errors.Wrap(err, "failed to validate message signature")
+	}
+
```
Comment (alonmuroch) [inline dkg/frost/frost.go:157]: do we check identifier, signature, etc?

File: dkg/frost/frost.go [hunk 8]
```diff
@@ -0,0 +1,756 @@
...
+	protocolMessage := &ProtocolMsg{}
+	if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+		return false, nil, errors.Wrap(err, "could not decode protocol msg")
+	}
+
+	if fr.msgs[protocolMessage.Round] == nil {
+		fr.msgs[protocolMessage.Round] = make(map[uint32]*dkg.SignedMessage)
+	}
+
+	originalMessage, ok := fr.msgs[protocolMessage.Round][uint32(msg.Signer)]
```
Comment (alonmuroch) [inline dkg/frost/frost.go:167]: why does this make it inconsistent? what if we simply get the same message twice?

File: dkg/frost/frost.go [hunk 9]
```diff
@@ -0,0 +1,756 @@
...
+			}
+			return true, out, nil
+		}
+	case Blame:
+		out, err := fr.processBlame()
+		if err != nil {
+			return false, nil, err
+		}
+		return true, &dkg.KeyGenOutput{BlameOutout: out}, err
+	}
```
Comment (alonmuroch) [inline dkg/frost/frost.go:207]: switch needs a default case

File: dkg/frost/frost.go [hunk 10]
```diff
@@ -0,0 +1,756 @@
...
+func (fr *FROST) canProceedRound1() bool {
+	if fr.currentRound != Preparation {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Preparation][operatorID]
+		if ok {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:518]: why is this tested here? Are we simply checking all operators sent a msg? if so please add some comments

File: dkg/frost/frost.go [hunk 11]
```diff
@@ -0,0 +1,756 @@
...
+		Round: Round1,
+		Round1Message: &Round1Message{
+			Commitment: commitments,
+			ProofS:     bCastMessage.Wi.Bytes(),
+			ProofR:     bCastMessage.Ci.Bytes(),
+			Shares:     shares,
+		},
+	}
+
+	protocolMessageBytes, err := protocolMessage.Encode()
```
Comment (alonmuroch) [inline dkg/frost/frost.go:254]: let's make a reusable broadcast func

File: dkg/frost/frost.go [hunk 12]
```diff
@@ -0,0 +1,756 @@
...
+func (fr *FROST) canProceedRound2() bool {
+	if fr.currentRound != Round1 {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Round1][operatorID]
+		if ok {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:542]: why are we decoding the msg? are we just checking all participants sent msg?

File: dkg/frost/frost.go [hunk 13]
```diff
@@ -0,0 +1,756 @@
...
+
+	protocolMessage := &ProtocolMsg{
+		Round: Round2,
+		Round2Message: &Round2Message{
+			Vk:      bCastMessage.VerificationKey.ToAffineCompressed(),
+			VkShare: bCastMessage.VkShare.ToAffineCompressed(),
+		},
+	}
+
+	protocolMessageBytes, err := protocolMessage.Encode()
```
Comment (alonmuroch) [inline dkg/frost/frost.go:352]: use reusable broadcasting func

File: dkg/frost/frost.go [hunk 14]
```diff
@@ -0,0 +1,756 @@
...
+	if fr.currentRound != Round2 {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Round2][operatorID]
+		if ok {
+			if err := protocolMessage.Decode(msg.Message.Data); err != nil {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:567]: why decoding? just checking all participants sent?

File: dkg/frost/frost.go [hunk 15]
```diff
@@ -0,0 +1,756 @@
...
+		out := &bls.G1{}
+		if err := bls.G1LagrangeInterpolation(out, xVec, yVec); err != nil {
+			return nil, err
+		}
+
+		outputs = append(outputs, out)
+	}
+
+	for i := 1; i < len(outputs); i++ {
+		fmt.Printf("vk: %x\n", outputs[i].Serialize())
```
Comment (alonmuroch) [inline dkg/frost/frost.go:730]: please remove

File: dkg/frost/frost.go [hunk 16]
```diff
@@ -0,0 +1,756 @@
...
+		blameOutput := &dkg.BlameOutput{
+			Valid:        valid,
+			BlameMessage: blameMessageBytes,
+		}
+		return blameOutput, nil
+	}
+	return nil, nil
+}
+
+func (fr *FROST) processBlameTypeInvalidShare(operatorID uint32, blameMessage *BlameMessage) (bool /*valid*/, error) {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:457]: maybe move all blame func to separate file for better and cleaner organization

File: dkg/frost/frost.go [hunk 17]
```diff
@@ -0,0 +1,756 @@
...
+		if err != nil {
+			return false, nil, err
+		}
+		return true, &dkg.KeyGenOutput{BlameOutout: out}, err
+	}
+
+	return false, nil, nil
+}
+
+func (fr *FROST) processRound1() error {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:212]: move maybe to its own file?

File: dkg/frost/frost.go [hunk 18]
```diff
@@ -0,0 +1,756 @@
...
+		return err
+	}
+
+	bcastRound1Message.Signature = sig
+	fr.msgs[Round1][uint32(fr.operatorID)] = bcastRound1Message
+
+	return fr.network.BroadcastDKGMessage(bcastRound1Message)
+}
+
+func (fr *FROST) processRound2() error {
```
Comment (alonmuroch) [inline dkg/frost/frost.go:288]: move to its own file?

File: go.mod [hunk 19]
```diff
@@ -4,27 +4,34 @@ go 1.17
...
 	github.com/prysmaticlabs/go-bitfield v0.0.0-20210809151128-385d8c5e3fb7
 	github.com/stretchr/testify v1.7.1
+	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
 )
 
 require (
+	filippo.io/edwards25519 v1.0.0-rc.1 // indirect
+	github.com/btcsuite/btcd v0.21.0-beta.0.20201114000516-e9c7a5ac6401 // indirect
 	github.com/btcsuite/btcd/btcec/v2 v2.2.0 // indirect
+	github.com/bwesterb/go-ristretto v1.2.0 // indirect
```
Comment (alonmuroch) [inline go.mod:23]: remove, not used

File: go.mod [hunk 20]
```diff
@@ -4,27 +4,34 @@ go 1.17
...
+	github.com/consensys/gnark-crypto v0.5.3 // indirect
```
Comment (alonmuroch) [inline go.mod:24]: not used

File: go.mod [hunk 21]
```diff
@@ -4,27 +4,34 @@ go 1.17
...
 	github.com/herumi/bls-eth-go-binary v0.0.0-20220509081320-2d8ab06de53c
 	github.com/pkg/errors v0.9.1
 	github.com/prysmaticlabs/go-bitfield v0.0.0-20210809151128-385d8c5e3fb7
 	github.com/stretchr/testify v1.7.1
+	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
 )
 
 require (
+	filippo.io/edwards25519 v1.0.0-rc.1 // indirect
+	github.com/btcsuite/btcd v0.21.0-beta.0.20201114000516-e9c7a5ac6401 // indirect
```
Comment (alonmuroch) [inline go.mod:21]: not used

File: go.mod [hunk 22]
```diff
@@ -4,27 +4,34 @@ go 1.17
...
+	github.com/ferranbt/fastssz v0.0.0-20220103083642-bc5fefefa28b
 	github.com/herumi/bls-eth-go-binary v0.0.0-20220509081320-2d8ab06de53c
 	github.com/pkg/errors v0.9.1
 	github.com/prysmaticlabs/go-bitfield v0.0.0-20210809151128-385d8c5e3fb7
 	github.com/stretchr/testify v1.7.1
+	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
 )
 
 require (
+	filippo.io/edwards25519 v1.0.0-rc.1 // indirect
```
Comment (alonmuroch) [inline go.mod:20]: not used

File: dkg/frost/frost_test.go [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+	for _, node := range mockNetwork.nodes {
+		go func() {
+			_ = node.Run()
+		}()
+		defer node.Exit()
+	}
+
+	threshold := 2
+
+	g := errgroup.Group{}
```
Comment (alonmuroch) [inline dkg/frost/frost_test.go:56]: we don't use async in testing, just have a bunch of msgs go to the frost implementation. Makes it easier to read

File: dkg/frost/mock_network.go [hunk 24]
```diff
@@ -0,0 +1,100 @@
+package frost
```
Comment (alonmuroch) [inline dkg/frost/mock_network.go:1]: we don't use mocks in implementation, that's why we have the testing utils

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: dkg/frost/mock_network.go (hunk 24 above)
Comment (gl-rockx) [reply dkg/frost/mock_network.go:1]: Will refactor the tests and use testing utils.

File: dkg/frost/frost.go (hunk 10 above)
Comment (aayush-rockx) [reply dkg/frost/frost.go:518]: yes, we are simply checking if we have received all messages from the previous round. Will add comments and extract reusable code from this

File: dkg/frost/frost.go (hunk 12 above)
Comment (aayush-rockx) [reply dkg/frost/frost.go:542]: We were decoding messages to check if there was an invalid message. I will remove this part as I've added this check before saving the message.

File: dkg/frost/frost.go (hunk 14 above)
Comment (aayush-rockx) [reply dkg/frost/frost.go:567]: yes, will remove this part as mentioned in previous comment

File: dkg/frost/frost.go (hunk 3 above)
Comment (aayush-rockx) [reply dkg/frost/frost.go:41]: It is to differentiate it from the default zero value of type int. If it starts from `0` i.e `iota` then `New` will set `currentRound` to `Preparation` which starts only after `Start` fn is called.

**Pull request merged**